### PR TITLE
Add pruning handler to read only replica

### PR DESCRIPTION
**Pull request by yontyon**
When pruning, we need to verify that read-only replicas already have all blocks until the pruning point. For this, we add a pruning handler to the read-only replica.
This pruning handler knows to handler only latestPruneableBlock request and it returns the lastReachableBlock of the read-only replica

### CODE REVIEW

File: bftengine/src/bftengine/ReadOnlyReplica.cpp
Comment (dartdart26) [inline bftengine/src/bftengine/ReadOnlyReplica.cpp:91]: Could you please explain the rationale for this change? I assume it is not related to this PR, but I might be wrong here.

File: bftengine/src/bftengine/ReadOnlyReplica.cpp
Comment (dartdart26) [inline bftengine/src/bftengine/ReadOnlyReplica.cpp:151]: Nit: Since we are in the `ClientRequestMsg` handler anyway, maybe it would be simpler to just do `spanContext<ClientRequestMsg>()`. Just a thought.

File: bftengine/src/bftengine/ReadOnlyReplica.cpp
Comment (dartdart26) [inline bftengine/src/bftengine/ReadOnlyReplica.cpp:162]: typo: reqeust -> request

File: bftengine/src/bftengine/ReadOnlyReplica.cpp
Comment (dartdart26) [inline bftengine/src/bftengine/ReadOnlyReplica.cpp:171]: Nit: I think we can use a `const ClientRequestMsg&` type as we know it can't be nullptr and we don't change it. Speaking of which, we might also change:

```
char* ClientRequestMsg::requestBuf() const;
```
to return `const char*`.

File: bftengine/src/bftengine/ReadOnlyReplica.cpp
Comment (dartdart26) [inline bftengine/src/bftengine/ReadOnlyReplica.cpp:177]: Nit: can be made `const`.

File: bftengine/src/bftengine/ReadOnlyReplica.hpp
Comment (dartdart26) [inline bftengine/src/bftengine/ReadOnlyReplica.hpp:44]: Could you please explain that change (as above)?

File: kvbc/include/pruning_handler.hpp
Comment (dartdart26) [inline kvbc/include/pruning_handler.hpp:195]: Nit: Maybe we want to say "... what the latest **prunable** block is".

File: kvbc/include/pruning_handler.hpp
Comment (dartdart26) [inline kvbc/include/pruning_handler.hpp:207]: I guess this might not be related to this PR, but I was wondering if we should return 0 as the last reachable block in case pruning is disabled. Does it make sense?

Comment (dartdart26) [reply kvbc/include/pruning_handler.hpp:207]: Actually, now that I think of it, maybe a read-only replica doesn't even care if pruning is enabled. What do you think?

File: util/pyclient/bft_client.py
Comment (dartdart26) [inline util/pyclient/bft_client.py:287]: Could you please explain the Apollo RSI changes?

File: kvbc/include/pruning_handler.hpp
Comment (dartdart26) [inline kvbc/include/pruning_handler.hpp:204]: Now the reconfiguration engine needs to verify the `LatestPrunableBlock` message from the read-only replica too. We need to change the `PublicKeys()` function in the pruning handler so that it includes read-only replicas, correct?

**Comment by dartdart26** [comment]
Looks great, thanks @yontyon! I have added a few minor comments and asked a couple of questions. In general, I think it is good to go.

File: bftengine/src/bftengine/ReadOnlyReplica.cpp
Comment (yontyon) [reply bftengine/src/bftengine/ReadOnlyReplica.cpp:91]: Previously we prevented any message from being processed while ST. now we want to allow client request and deny any other message (Checkpoint message)

File: bftengine/src/bftengine/ReadOnlyReplica.hpp
Comment (yontyon) [reply bftengine/src/bftengine/ReadOnlyReplica.hpp:44]: We want to allow client requests (because the replica is mostly on ST and we want to allow the latestPruneableBlock request even then), but still to deny checkpoint messages

File: kvbc/include/pruning_handler.hpp
Comment (yontyon) [reply kvbc/include/pruning_handler.hpp:207]: We want not to touch anything in case pruning is disabled, because there might be other handlers that will be called (currently it is only concord's pruning handler, we want only one pruning handler (concord's) and for that we simply deactivate the internal command handler)

File: util/pyclient/bft_client.py
Comment (yontyon) [reply util/pyclient/bft_client.py:287]: Currently, the common data contain also the primary. However, we want to enable read requests to read only replicas, and ro replica doesn't know who the primary is

File: kvbc/include/pruning_handler.hpp
Comment (yontyon) [reply kvbc/include/pruning_handler.hpp:204]: yes

File: bftengine/src/bftengine/ReadOnlyReplica.cpp
Comment (dartdart26) [reply bftengine/src/bftengine/ReadOnlyReplica.cpp:91]: Thanks!

File: kvbc/include/pruning_handler.hpp
Comment (dartdart26) [reply kvbc/include/pruning_handler.hpp:207]: I see, thanks. I was thinking that we might want to always return the last block for read-only replica, irrespective of `pruning_enabled_`.

Also, wanted to ask something else. What will the value be for `latest_prunable_block` if `pruning_enabled_ = false`?

File: util/pyclient/bft_client.py
Comment (dartdart26) [reply util/pyclient/bft_client.py:287]: Does the primary now become part of RSI?

**Review by dartdart26 — APPROVED** [review]
Thanks for the explanations @yontyon!

File: util/pyclient/bft_client.py
Comment (andrewjstone) [reply util/pyclient/bft_client.py:287]: I'm not sure this change is correct. The knowledge of the primary is required for normal operation of the client. The client should only send to the primary for write operations, unless there is a timeout. Sending to all replicas should only be done on timeout to ensure a primary is not malicious, otherwise we are sending more traffic than necessary. It shouldn't induce view change, but it's still different from the behavior of the real C++ client, and doesn't seem right to me. 

Excluding the primary is probably fine for read-only operations, but not write operations.

Comment (yontyon) [reply util/pyclient/bft_client.py:287]: @andrewjstone, the replicas still send the write requests to the primary only. This is just an implementation change. We collect the results from the replicas for a specific sequence number under a common key (as the reply may contain RSI). This change means that the primary number is not part of this common data, but it is still part of the reply itself, once the client decides the replies can be delivered up, it delivers also the primary ID.
We removed it from the common data, because the RORs don't have the knowledge on who the primary is, and we would like to collect their responses together with the committers reply.

File: kvbc/include/pruning_handler.hpp
Comment (yontyon) [reply kvbc/include/pruning_handler.hpp:207]: The point is that we may want to return something else from another pruning handler. So if the basic pruning (or any other pruning handler) is not enabled, we want not to touch the latest pruneable block as we may override a previous value of it. The value will be the one that set to the reference by either another pruning handler or by the dispatcher.

File: util/pyclient/bft_client.py
Comment (yontyon) [reply util/pyclient/bft_client.py]: @andrewjstone in any case, I revreted this change. I will perform vast changes once I will merge the Apollo tests for this

**Pull request merged**
